### PR TITLE
Pending intent

### PR DESCRIPTION
**Pull request by sarveshkapre**
**Description**

Added an extra check to detect use of empty pending intents.

### CODE REVIEW

File: qark/modules/findPending.py [hunk 1]
```diff
@@ -21,456 +22,510 @@
...
 method_list.append('getActivities')
 method_list.append('getService')
 method_list.append('getActivity')
 method_list.append('getBroadcast')
 
 
-def start(queue,height):
-	"""
+def start(queue, height):
+    """
```
Comment (neilatli) [inline qark/modules/findPending.py]: This ends up looking like 
```
    """
	Start finding pending intents
	"""

File: qark/modules/findPending.py [hunk 2]
```diff
@@ -21,456 +22,510 @@
...
+
+                                issue = terminalPrint()
+                                issue.setLevel(Severity.WARNING)
+                                issue.setData("Empty Pending instance found in: " + str(
+                                    filename) + ". For security reasons, the Intent you supply here should almost always be an explicit intent that is specify an explicit component to be delivered to through Intent.setClass. A malicious application could potentially intercept, redirect and/or modify this Intent. Pending Intents retain the UID of your application and all related permissions, allowing another application to act as yours. Reference: https://developer.android.com/reference/android/app/PendingIntent.html")
+                                results.append(issue)
+    return
+
+def recurse(f, t, results):
+    """
```
Comment (neilatli) [inline qark/modules/findPending.py]: Same as above.

File: qark/modules/findPending.py [hunk 3]
```diff
@@ -21,456 +22,499 @@
...
+    """ Start finding pending intents """
+    results = []
+    global tree
+    global current_file
+    count = 0
+
+    # TODO - Look for use of fillIn method which can make this a much more exploitable condition
+    for j in common.java_files:
+        count = count + 1
+        pub.sendMessage('progress', bar='Pending Intents', percent=round(count * 100 / common.java_files.__len__()))
```
Comment (dhaugh) [inline qark/modules/findPending.py]: Not crucial, but I'm pretty sure you can use `len(common.java_files)` here

File: qark/modules/findPending.py [hunk 4]
```diff
@@ -21,456 +22,499 @@
...
+        current_file = j
```
Comment (dhaugh) [inline qark/modules/findPending.py]: this seems extraneous
I think you could just do `for current_file in common.java_files`, which will lead to less variables floating around.

File: qark/modules/findPending.py [hunk 5]
```diff
@@ -21,456 +22,499 @@
...
+            common.logger.debug("No type declarations: " + str(j))
+            report.write("parsing error-issues-list", str(current_file), "strong")
+    queue.put(results)
+    return
+
+
+def empty_intent(t, filename, results):
+    if type(t) is m.MethodDeclaration:
+        for x in method_list:
+            if str(x) in str(t):
```
Comment (dhaugh) [inline qark/modules/findPending.py]: it's kind of hard to diffrentiate what x and t mean here
would recommend renaming x to method_obj (or something similar)  and t to method_declaration. Would also rename the function to something like `check_method_declaration_for_intents`, and mention in a comment somewhere that it modifies the result list (I think it does anyway?)

File: qark/modules/findPending.py [hunk 6]
```diff
@@ -21,456 +22,499 @@
...
+                            for t in type_decl.body:
+                                try:
+                                    empty_intent(t, j, results)
+                                except Exception as e:
+                                    common.logger.debug("Problem in empty intent function of findPending.py: " + str(e))
+                                    common.parsingerrors.add(str(current_file))
+                                for f in t._fields:
+                                    # dynamically parse each token where is f is the field and t is the token
+                                    try:
+                                        recurse(f, t, results)
```
Comment (dhaugh) [inline qark/modules/findPending.py]: It's non-obvious what this function does. Rename it to something like `recursive_search_for_<whatever you're looking for>` at least enough for context

Comment (dhaugh) [inline qark/modules/findPending.py]: Also mention that it adds to results (either inline comment or function name)

File: qark/modules/findPending.py [hunk 7]
```diff
@@ -21,456 +22,499 @@
...
+                                        parse_args(current.arguments[2], results)
+                                    except Exception as e:
+                                        report.write("parsingerror-issues-list", str(current_file), "strong")
+                                        common.logger.debug(
+                                            "Problem in parse_args function of findPending.py: " + str(e))
+            elif l is None:
+                pass
+            else:
+                if type(l) is not str:
+                    for n in l._fields:
```
Comment (dhaugh) [inline qark/modules/findPending.py]: Are you positive that l._fields is iterable without knowing the type?

File: qark/modules/findPending.py [hunk 8]
```diff
@@ -21,456 +22,499 @@
...
+
+                                issue = terminalPrint()
+                                issue.setLevel(Severity.WARNING)
+                                issue.setData("Empty Pending instance found in: " + str(
+                                    filename) + ". For security reasons, the Intent you supply here should almost always be an explicit intent that is specify an explicit component to be delivered to through Intent.setClass. A malicious application could potentially intercept, redirect and/or modify this Intent. Pending Intents retain the UID of your application and all related permissions, allowing another application to act as yours. Reference: https://developer.android.com/reference/android/app/PendingIntent.html")
+                                results.append(issue)
+    return
+
+
+def recurse(f, t, results):
```
Comment (dhaugh) [inline qark/modules/findPending.py]: I'd write a bunch of helper functions for this one. It looks like at a high level, it looks like a switch statement on `type(current)`. Perhaps a separate helper function for each type?

File: qark/modules/findPending.py [hunk 9]
```diff
@@ -21,456 +22,499 @@
...
+                if recurse_class(classname, type_decl, results):
+                    return True
+    return found
+
+
+def recurse_class(classname, treeObject, results):
+    found = False
+    if hasattr(treeObject, '_fields'):
+        for x in treeObject._fields:
+            if type(getattr(treeObject, x)) is list:
```
Comment (dhaugh) [inline qark/modules/findPending.py]: like above, this looks like the start of a switch statement where each function can be broken out into a seperate helper function

File: qark/modules/findPending.py [hunk 10]
```diff
@@ -21,456 +22,499 @@
...
+
+def intent_digger(t, intent):
+    """ Traverse the intent declaration hierarchy """
+    results = [False, False]
+    if hasattr(t, "_fields"):
+        for f in t._fields:
+            current = getattr(t, f)
+            if type(current) is list:
+                for c in current:
+                    if type(c) is m.VariableDeclaration:
```
Comment (dhaugh) [inline qark/modules/findPending.py]: Multiple nested if statements can be replaced with a single `if ___ and ___...`

Comment (dhaugh) [inline qark/modules/findPending.py]: In general, keep in mind how many levels deep your function is going (indentation is a good measure)
if you're 8 levels in, it's going to be very difficult to hold everything in your memory, and will be more bug-prone and harder to reason about

**Pull request merged**